### ISSUE TITLE
Crossplane init container does not wait for CRDs

### ISSUE THREAD

**Issue by sttts** [crossplane/crossplane#5828]
### What happened?

https://github.com/crossplane/crossplane/blob/master/internal/initializer/crds.go has not code to wait for the CRDs to be established and served. The result is a race between further init steps and kube-apiserver that can lead to:

```
          uxp │ [crossplane-init] ***"level":"info","ts":"2024-07-12T13:13:44Z","logger":"crossplane","msg":"Step has been completed","Name":"CoreCRDs"***
          uxp │ [crossplane-init] ***"level":"info","ts":"2024-07-12T13:13:44Z","logger":"crossplane","msg":"Step has been completed","Name":"WebhookConfigurations"***
          uxp │ [crossplane-init] crossplane: error: cannot initialize core: cannot apply lock object: cannot get object: no matches for kind "Lock" in version "pkg.crossplane.io/v1beta1"
```

### What environment did it happen in?
Crossplane version:

**Comment by jbw976** [comment]
Thanks for reporting this @sttts! What further consequences did you observe? Was it just that log statement and things recovered OK after that, did it cause a pod crash and restart, did it not recover at all?

**Comment by sttts** [comment]
`helm install` fails with that message.

**Comment by sailendhran** [comment]
@jbw976 Can I look into this issue?  Could you please assign it to me?

**Comment by jbw976** [comment]
Sure @sailendhran, that would be awesome! thank you for your initiative here 🙇‍♂️

**Comment by PrateekKumar1709** [comment]
@jbw976 I would like to work on this issue. Can you please assign this to me. Thanks.

**Comment by Rajan-226** [comment]
@jbw976 @sttts I would like to work and close this issue.

Upon investigation, it appears that this issue is one of the rare occurrences, as I was unable to reproduce it by simply running helm install multiple times.

I believe the root cause is that after applying a CRD, there is a delay before the RESTful API endpoint for the CRD is registered. [Reference](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#:~:text=When%20you%20create%20a%20new,scope%20field.)

I reviewed the code and suggest adding a timer and ticker [here](https://github.com/crossplane/crossplane/blob/1b4de654d10ad46f7c3058510c1a28f8981953df/internal/initializer/crds.go#L134) to wait after applying the CRDs and check for the status.conditions.type of the CRDs to be Established or just try to get the CRD with delay. With this approach, we could also remove [this step](https://github.com/crossplane/crossplane/blob/1b4de654d10ad46f7c3058510c1a28f8981953df/cmd/crossplane/core/init.go#L106) which seems to have been added for the [same purpose](https://github.com/crossplane/crossplane/pull/2160/commits/4b24cf7378b985c446710614c0914189e209b33e#diff-9cb8e0c3a364462ebfd576f489279834a684fa459dec0b841f009a6afc86800f). 

Please let me know if this approach aligns with your thoughts so that I can begin working on it.

**Comment by sttts** [comment]
yes, that sounds like the right direction, to wait for `Established`.

**Comment by Rajan-226** [comment]
@sttts @jbw976 Can someone review https://github.com/crossplane/crossplane/pull/5949?

**Comment by sailendhran** [comment]
Sorry folks. I was not able to make progress on this. Seems like @Rajan-226 has PR in review. I'll unassign myself from this issue. Following the PR for learning.
